cpp/include/cudf/detail/iterator.cuh

```diff
@@ -80,6 +80,34 @@ struct null_replaced_value_accessor
   }
 };
 
+/** -------------------------------------------------------------------------*
+ * @brief validity accessor of column with null bitmask
+ * A unary functor returns validity at `id`.
+ * `operator() (cudf::size_type id)` computes validity flag at `id`
+ * This functor is only allowed for nullable columns.
+ *
+ * @throws `cudf::logic_error` if the column is not nullable.
+ * -------------------------------------------------------------------------**/
+struct validity_accessor {
+  column_device_view const col;
+
+  /** -------------------------------------------------------------------------*
+   * @brief constructor
+   * @param[in] _col column device view of cudf column
+   * -------------------------------------------------------------------------**/
+  validity_accessor(column_device_view const& _col)
+    : col{_col}
+  {
+    // verify valid is non-null, otherwise, is_valid() will crash
+    CUDF_EXPECTS(_col.nullable(), "Unexpected non-nullable column.");
```

Comment:
If we allow for both nullable/non-nullable columns with this iterator, then we have to use the slower `is_valid` instead of `is_valid_nocheck()`.

I would argue that if the column is not nullable, then a separate code path should be used:
```
if(c.has_nulls){
   auto iter = make_validity_iterator(c);
} else {
   auto iter = thrust::constant_iterator<bool>(true);
}
```